### PR TITLE
Release/0.6.0

### PR DESCRIPTION
**Pull request by chuwy**
- [x] Update README usage example
- [x] Update remaining headers to 2019
- [x] [Fix AST conversion](https://github.com/snowplow/iglu-scala-client/blob/release/0.6.0/src/main/scala/com.snowplowanalytics.iglu/client/validator/CirceValidator.scala#L86-L87)

### CODE REVIEW

**Comment by chuwy** [comment]
Hey @BenFradet! This is an initial version of refactored Iglu Client. I think likely it will have even more changes, but you already can have a look.

File: src/main/scala/com.snowplowanalytics.iglu/client/ClientError.scala [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import validator.ValidatorReport
+import resolver.LookupHistory
+import resolver.registries.RegistryError
+
+sealed trait ClientError
+
+object ClientError {
+
+  /** Error happened during schema resolution step */
+  case class ResolverError(value: Map[String, LookupHistory]) extends ClientError {
```
[the remark this answers is not in the document]
Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/ClientError.scala]: Certainly!

File: src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+// This project
+import com.snowplowanalytics.iglu.client.ClientError.ResolverError
+import com.snowplowanalytics.iglu.client.resolver.registries.{
+  Registry,
+  RegistryError,
+  RegistryLookup
+}
+
+/** Resolves schemas from one or more Iglu schema registries */
+case class Resolver[F[_]](cache: Option[SchemaCache[F]], repos: List[Registry]) {
```
[the remark this answers is not in the document]
Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala]: Yep. Artifact from a previous version.

File: src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+   *                  this schema in Iglu
+   * @param remaining A List of repositories we have to look in
+   *                  (not-tried yet or with non-404 error)
+   * @param tried A Map of repositories with their accumulated errors
+   *              we have looked in fruitlessly so far
+   * @return either a Success-boxed schema (as a Json),
+   *         or a Failure-boxing of Map of repositories with all their
+   *         accumulated errors
+   */
+  def traverseRepos[F[_]: Monad: RegistryLookup](
```
[the remark this answers is not in the document]
Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala]: Yep, will add if its actually tail recursive.

File: src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+   *                 retries will happen in *next* requests
+   * @return a Validation boxing either the Schema's
+   *         Json on Success, or an error String
+   *         on Failure
+   */
+  def lookupSchema(schemaKey: SchemaKey, attempts: Int)(
+    implicit
+    F: Monad[F],
+    L: RegistryLookup[F],
+    C: Clock[F]): F[Either[ResolverError, Json]] = {
```
[the remark this answers is not in the document]
Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala]: Not sure. Do you mean something like `MonadClock[F[_]] extends Monad[F] with Clock[F]`?

File: src/main/scala/com.snowplowanalytics.iglu/client/resolver/SchemaCache.scala [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+   *
+   * @param storedTstamp timestamp when value was saved
+   * @return true if
+   */
+  private def isViable(unixSeconds: Int, storedTstamp: Int): Boolean =
+    ttl.fold(true)(ttlVal => unixSeconds - storedTstamp < ttlVal)
+}
+
+object SchemaCache {
+  def init[F[_]: Applicative: Cache](size: Int, ttl: Option[Int]): F[Option[SchemaCache[F]]] =
```
[the remark this answers is not in the document]
Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/SchemaCache.scala]: Tbh, I'm more and more leaning towards scalacache for this purpuse.

File: src/main/scala/com.snowplowanalytics.iglu/client/resolver/SchemaCache.scala [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    Option(())
+      .filter(_ => size > 0)
+      .filter(_ => !ttl.exists(_ <= 0)) match {
```
[the remark this answers is not in the document]
Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/SchemaCache.scala]: https://github.com/snowplow/iglu-scala-client/pull/105#discussion_r215933308

File: src/main/scala/com.snowplowanalytics.iglu/client/resolver/package.scala [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+  /** Schema lookup result associated with timestamp (in seconds) it was stored at */
+  type SchemaLookupStamped = (Int, SchemaLookup)
+
+  /** Our List (possibly empty) of Iglu repositories */
+  type RepositoryRefs = List[Registry]
+
+  type Result = Either[RegistryError, Json]
+
+  type ResolveResult = Either[ResolverError, Json]
```
[the remark this answers is not in the document]
Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/package.scala]: No, they're not. Also artifacts from old client.

File: src/main/scala/com.snowplowanalytics.iglu/client/resolver/registries/Utils.scala [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+      URI.create(url).asRight
+    } catch {
+      case _: NullPointerException =>
+        RegistryError.ClientFailure("Provided URL was null").asLeft
+      case e: IllegalArgumentException =>
+        val error = ExceptionUtils.getRootCause(e).getMessage
+        RegistryError.ClientFailure(s"Provided URI string violates RFC 2396: [$error]").asLeft
+    }
+
+  implicit val uriDecoder: Decoder[URI] =
```
[the remark this answers is not in the document]
Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/registries/Utils.scala]: Nope, surprisingly.

File: src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (BenFradet) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala]: no I meant `case class Resolver[F[_]: Monad: Clock]`

File: src/main/scala/com.snowplowanalytics.iglu/client/resolver/SchemaCache.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (BenFradet) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/SchemaCache.scala]: iirc we already investigated scalacache and it didn't fit our purposes?

File: src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala]: Ah no. It's actually a common tagless pattern to never restrict the data definition and each function can use its least poweful typeclass. I also remember reading an explanation on /r/haskell that one should never do that, but cannot find that reasoning right now.

[the remark this answers is not in the document]
Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala]: Probably this reasoning: https://stackoverflow.com/questions/24465586/why-constraints-on-data-are-a-bad-thing (too lazy to read through whole answer right now)

File: src/main/scala/com.snowplowanalytics.iglu/client/resolver/SchemaCache.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/SchemaCache.scala]: Yeah, this was the discussion https://github.com/snowplow/scala-maxmind-iplookups/issues/42. I think main reason was that its quite overloaded, but I'm not sure anymore that this is the price we shouldn't pay. I'm open for another discussion.

[the remark this answers is not in the document]
Comment (BenFradet) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/SchemaCache.scala]: I think two issues still remain: no lru implementation and limited options for in-memory (guava being most likely out of the question).

[the remark this answers is not in the document]
Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/SchemaCache.scala]: OHC is in-memory backend that has LRU as well as LFU: https://github.com/snazy/ohc#eviction-algorithms

[the remark this answers is not in the document]
Comment (BenFradet) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/SchemaCache.scala]: problem is that ohc is off-heap

[the remark this answers is not in the document]
Comment (BenFradet) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/SchemaCache.scala]: that might not play nice with spark given that it has its own off-heap mechanism

[the remark this answers is not in the document]
Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/SchemaCache.scala]: Okay, I didn't consider off-heap can be a problem.

File: src/main/scala/com.snowplowanalytics.iglu/client/ClientError.scala [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.iglu.client
+
+// cats
+import validator.ValidatorError
+import resolver.LookupHistory
+import resolver.registries.RegistryError
+
+sealed trait ClientError
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.iglu/client/ClientError.scala]: I think we need to extend exception

Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/ClientError.scala]: Done in M4. I also added `Encoder`/`Decoder` instances for all entities.

File: src/main/scala/com.snowplowanalytics.iglu/client/resolver/registries/RegistryLookup.scala [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+      def lookup(repositoryRef: Registry, schemaKey: SchemaKey): F[Either[RegistryError, Json]] =
+        repositoryRef match {
+          case Registry.Http(_, connection)  => httpLookup(connection, schemaKey)
+          case Registry.Embedded(_, path)    => embeddedLookup[F](path, schemaKey)
+          case Registry.InMemory(_, schemas) => F.pure(inMemoryLookup(schemas, schemaKey))
+        }
+    }
+
+  // Id instance also swallows all exceptions into `RegistryError`
+  implicit def idLookupInstance: RegistryLookup[Id] = new RegistryLookup[Id] {
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.iglu/client/resolver/registries/RegistryLookup.scala]: could we have a new one with `Eval`?

Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/registries/RegistryLookup.scala]: Yep, will switch everything to `Eval` today

Comment (chuwy) [reply src/main/scala/com.snowplowanalytics.iglu/client/resolver/registries/RegistryLookup.scala]: @BenFradet, `M5` with `RegistryLookup[Eval]` will be on its way to Maven after `scala-lru-map:0.3.0-M3` is published.

**Pull request merged**